addons/product/models/product_template.py

```diff
@@ -938,6 +939,61 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
         """
         return next(self._get_possible_combinations(parent_combination, necessary_values), self.env['product.template.attribute.value'])
 
+    def _cartesian_product(self, attribute_values):
+        """
+        Generate all possible combination for attributes values.
+        It is equivalent to itertools.product except it skips invalid combinations.
+
+        It's necessary for performance reason because filtering out invalid combinations from standard Cartesian product
+        can be extremely slow
+        """
+        if not attribute_values:
+            raise StopIteration
+
+        pools = attribute_values
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        # a value is excluded if its counter is greater than 0
+        current_exclusions = defaultdict(int)
+
+        values_idx = [-1] * len(pools)
+        pool_idx = 0
+
+        def current_value_id():
+            return pools[pool_idx][values_idx[pool_idx]].id
+
+        while True:
+            if values_idx[pool_idx] >= 0:
+                # remove exclusions of current value from current combination
```

Comment:
These comments are just covering up bad variable names.

Good comments would be like what you explained me the other day, for example that the goal is to immediately discard all subsets of combination that include at least 2 know values that are imcompatibles.

Reply:
This part of the code is not about skipping. The comment was here because `-= 1` isn't obvious and yet it's necessary. I'll rewrite the comment and move it closer to the statement.